Objects/weakrefobject.c

```diff
@@ -276,64 +426,157 @@ insert_head(PyWeakReference *newref, PyWeakReference **list)
     *list = newref;
 }
 
+/* See if we can reuse either the basic ref or proxy instead of creating a new
+ * weakref
+ */
+static PyWeakReference *
+try_reuse_basic_ref(PyWeakReference *ref, PyWeakReference *proxy,
+                    PyTypeObject *type, PyObject *callback)
+{
+    if (callback != NULL) {
+        return NULL;
+    }
+
+    PyWeakReference *cand = NULL;
+    if (type == &_PyWeakref_RefType) {
+        cand = ref;
+    }
+    if (type == &_PyWeakref_ProxyType) {
+        cand = proxy;
+    }
```

Comment:
Doesn't this also need to handle `_PyWeakref_CallableProxyType`?

Reply:
Yep, good catch!